Framework/PythonInterface/plugins/algorithms/SetSampleFromLogs.py

```diff
@@ -0,0 +1,167 @@
+from mantid.api import (AlgorithmFactory, DistributedDataProcessorAlgorithm)
+from mantid.kernel import (ConfigService)
+from mantid import _kernel as mtd_kernel
+from mantid.simpleapi import (SetSample)
+
+
+PROPS_FOR_SETSAMPLE = ["InputWorkspace", "Geometry", "Material", "Environment", "ContainerGeometry", "ContainerMaterial"]
+
+
+def _toDict(propertyManager):
+    result = {}
+    if propertyManager:
+        for key in propertyManager.keys():
+            result[key] = propertyManager[key].value
+            if isinstance(result[key], mtd_kernel.std_vector_dbl):
+                result[key] = list(result[key])
+    return result
+
+
+def _findKey(dictlike, *args):
+    for name in args:
+        if (name in dictlike) and bool(dictlike[name]):
+            return name
+    return ''  # indicates failure
+
+
+def _hasValue(dictLike, key):  # TODO refactor into *args and return the one that exists or empty string
+    return (key in dictLike) and bool(dictLike[key])
+
+
+def _getLogValue(propertyManager, key):
+    if not _hasValue(propertyManager, key):
+        raise ValueError('Failed to find "{}" in logs'.format(key))
+    return propertyManager[key].lastValue()
+
+
+class SetSampleFromLogs(DistributedDataProcessorAlgorithm):
+    def category(self):
+        return "Sample"
+
+    def seeAlso(self):
+        return ["SetSample"]
+
+    def name(self):
+        return "SetSampleFromLogs"
+
+    def summary(self):
+        return "This algorithm looks through the logs to automatically determine the sample geometry and material"
+
+    def PyInit(self):
+        self.copyProperties("SetSample", PROPS_FOR_SETSAMPLE)
+        self.declareProperty("FindGeometry", True,
+                             "Whether to look for the 'Height' parameter in the logs")
+        self.declareProperty("FindSample", True,
+                             "Whether to look for the sample material in the logs")
+        self.declareProperty("FindEnvironment", True,
+                             "Whether to look for the sample container in the logs")
+
+    def PyExec(self):
+        wksp = self.getProperty("InputWorkspace").value
+        geometry = _toDict(self.getProperty("Geometry").value)
+        material = _toDict(self.getProperty("Material").value)
+        environment = _toDict(self.getProperty("Environment").value)
+        geometryContainer = _toDict(self.getProperty("ContainerGeometry").value)
+        materialContainer = _toDict(self.getProperty("ContainerMaterial").value)
+        findGeometry = self.getProperty("FindGeometry").value
+        findSample = self.getProperty("FindSample").value
+        findEnvironment = self.getProperty("FindEnvironment").value
+
+        # get a convenient handle to the logs
+        runObject = wksp.run()
+
+        # get information for the material from the logs
+        if findSample:
+            if (not _hasValue(material, 'ChemicalFormula')) and _hasValue(runObject, "SampleFormula"):
```

Comment:
My understanding is that Python dictionary already support this feature. For example, we can directly use
`dict.get("keyname", default_value)`.

Reply:
The problem is if `runObject` doesn't have a "SampleFormula" log, there is no default value to grab. Also, if the user already specified `material['ChemicalFormula']` we should just leave the value there. I don't like the verbosity of the `if` either and welcome a more compact way to specify it.